cert/cert-renewal/src/policy.rs

```diff
@@ -0,0 +1,361 @@
+    pub threshold: Policy,
+    pub retry: Policy,
+}
+
+/// Determines the policy for certificate renewal and retries.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Policy {
+    /// Renew and retry as a percentage of the certificate's lifetime.
+    /// This value is always between 0 and 100.
+    Percentage(i64),
+
+    /// Renew and retry at fixed time intervals around expiry.
+    Time(i64),
+}
+
+impl Policy {
+    /// Check if the given cert should be renewed based on this policy.
+    pub fn should_renew(&self, cert: &openssl::x509::X509) -> bool {
+        let not_before = cert.not_before();
+        let not_after = cert.not_after();
+        let now = now();
+
+        // Cert not_after must be after not_before to be valid.
+        // Always renew invalid certs.
+        if not_before >= not_after {
+            return true;
+        }
+
+        // Check for expired certificate. Renew if expired.
+        if now >= not_after {
+            return true;
+        }
+
+        match self {
+            Policy::Percentage(threshold) => {
+                let total_lifetime = not_before.diff(not_after).expect("valid times should diff");
+                let total_lifetime = timediff_to_secs(&total_lifetime);
+
+                let current_lifetime = not_before.diff(&now).expect("valid times should diff");
+                let current_lifetime = timediff_to_secs(&current_lifetime);
+
+                let lifetime = current_lifetime * 100 / total_lifetime;
+
+                lifetime >= *threshold
+            }
+
+            Policy::Time(threshold) => {
+                let expiry = asn1time_to_unix(not_after);
+                let now = asn1time_to_unix(&now);
+
+                expiry - now <= *threshold
+            }
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for Policy {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let mut policy: String = serde::Deserialize::deserialize(deserializer)?;
+
+        let unit = if let Some(unit) = policy.pop() {
+            unit
+        } else {
+            return Err(serde::de::Error::custom("policy not specified"));
+        };
+
+        let value: i64 = if let Ok(value) = policy.parse() {
+            value
+        } else {
+            return Err(serde::de::Error::custom(
+                "policy could not be parsed as number",
+            ));
+        };
+
+        if value <= 0 {
+            return Err(serde::de::Error::custom(
+                "renewal policy values must be positive",
+            ));
+        }
+
+        match unit {
+            '%' => {
+                if value >= 100 {
+                    return Err(serde::de::Error::custom(
+                        "lifetime percentage must be less than 100",
+                    ));
+                }
+
+                Ok(Policy::Percentage(value))
+            }
+            'm' => Ok(Policy::Time(value * 60)),
```

Comment:
I wonder if people will think "m" means "months" instead of "minutes", given that the other option is "days".

Maybe we should require them in full? "day", "min"

Reply:
Changed to "day" and "min".